### ISSUE TITLE
[Request]: Lock timeline scroll in Federated view when not at top

### ISSUE THREAD

**Issue by atirut-w** [GeopJr/Tuba#1045]
### Describe the request

From what I see, Tuba currently does not "lock" the scroll when you scroll down the Federated timeline and instead just stop showing new posts. This gets a little annoying when you're near the top enough that the app doesn't stop the flow of posts. See attached video for example.

https://github.com/GeopJr/Tuba/assets/25323231/fa15b610-8e88-494a-bd47-d58a0a1b761a

### Implementation Details

- [X] This follows the [GNOME HIG](https://developer.gnome.org/hig/).

**Comment by GeopJr** [comment]
Thanks for raising this issue!

> From what I see, Tuba currently does not "lock" the scroll when you scroll down the Federated timeline and instead just stop showing new posts.

There's no such thing as 'locking' the scroll view. When new items get appended, the scroll position stays the same but the content 'moves down'. To overcome that, Tuba has a queue. When you are not near the top, new posts get added to the queue and when u go near the top again, those posts get added to the view. That also matches the mastodon-web behavior somewhat, where if you are at the top, new posts get shown/prepended.

> This gets a little annoying when you're near the top enough that the app doesn't stop the flow of posts. See attached video for example.

I understand that the federated timeline gets updated constantly and the 'near the top' behavior is not good enough, maybe limiting it to only prepend the posts when you are at the absolute top then?

**Comment by atirut-w** [comment]
> When new items get appended, the scroll position stays the same but the content 'moves down'.

Would it be possible to keep the content in place as new contents are added instead?

> I understand that the federated timeline gets updated constantly and the 'near the top' behavior is not good enough, maybe limiting it to only prepend the posts when you are at the absolute top then?

This could work better than the current implementation, but I'm not a UI nor UX designer by any means.

**Comment by kimothy** [comment]
> Thanks for raising this issue!
> 
> > From what I see, Tuba currently does not "lock" the scroll when you scroll down the Federated timeline and instead just stop showing new posts.
> 
> There's no such thing as 'locking' the scroll view. When new items get appended, the scroll position stays the same but the content 'moves down'. To overcome that, Tuba has a queue. When you are not near the top, new posts get added to the queue and when u go near the top again, those posts get added to the view. That also matches the mastodon-web behavior somewhat, where if you are at the top, new posts get shown/prepended.
> 
> > This gets a little annoying when you're near the top enough that the app doesn't stop the flow of posts. See attached video for example.
> 
> I understand that the federated timeline gets updated constantly and the 'near the top' behavior is not good enough, maybe limiting it to only prepend the posts when you are at the absolute top then?

Hi! I waned to thank you for the app, it´s really well done, I use it every day!

I also wanted to come with a proposal for an improvement, and I found this ticket.

When reading your comment I learned that if I scrolled far enough down, then the scroll view "locks", which is the behavior that I would like to have. The issue that I have is that when I look at the federated posts and something catches my eye, if I scroll down as much as is needed, the post that was interesting is not really possible to see, and if you scroll back up again, it all starts flowing again. If this makes sense?

What i would like to suggest is that as soon as you scroll a little bit down, or click on a post or on the timeline, the the feed stops and you can scroll up and down as much as you like. If you scroll all the way back to the top, and stay there for a second or two, it starts to flow. Maybe have an indicator that shows that it will start flowing again.

This is my proposal ideas to improve the view.

Have a nice weekend, and thanks a lot for you hard work!

**Comment by Baralheia** [comment]
I'd second something like this for all timelines in Tuba, tbh... I'd prefer it if each timeline only automatically updated when you are at the absolute top of the timeline. Ideally, the timeline should freeze if you scroll down even just a little bit. If new posts come in while the timeline is frozen, there should be some sort of indicator that there are new posts to view. Perhaps the "Scroll to top" button could either change color or have a star/dot/etc type overlay to show that there's new posts waiting to be seen?

**Comment by GeopJr** [comment]
> Would it be possible to keep the content in place as new contents are added instead?

No, this is another thing ListView would fix but [insert list of frustrations with migrating to them]. Technically would be possible by calculating heights and re-calculating the scrolled position but when you get to the details it's almost impossible to do right in a speedy way.

> This could work better than the current implementation, but I'm not a UI nor UX designer by any means.

> This is my proposal ideas to improve the view.

The main problem IIRC was that people were *very* close to the top but not exactly at the top and wondered why the queue didn't prepend. I guess lowering the threshold to something very small could work.

> If new posts come in while the timeline is frozen, there should be some sort of indicator

FWIW, there is, there will be a non-prominent badge either at the tab on narrow mode or the sidebar entry in wide mode.
